### PR TITLE
chore(example): update example

### PR DESCRIPTION
**Pull request by fibonacci1729**
~~I'm not sure that using `wasm32-unknown-unknown` is actually the best solution to get this working but I assume we can't use `wasm32-wasi` without doing something fancier with preview2-prototyping?~~

~~In any case, these are the changes I had to make to get the compose example working.~~

Updated to use latest version of `cargo-component` and compatible adapter.

### CODE REVIEW

**Comment by fibonacci1729** [comment]
cc @peterhuene

**Comment by peterhuene** [comment]
The examples should work with the wasi target, so let me look into it; probably just something out of date.

**Comment by fibonacci1729** [comment]
It seems like various wasi imports need to be satisfied by the server when linking the component:

```
❯ wasm-tools compose -c config.yml -o service.wasm ../middleware/target/wasm32-wasi/release/middleware.wasm
[2023-04-18T16:31:10Z WARN ] instance `streams` will be imported because a dependency named `streams` could not be found
[2023-04-18T16:31:10Z WARN ] instance `filesystem` will be imported because a dependency named `filesystem` could not be found
[2023-04-18T16:31:10Z WARN ] instance `environment` will be imported because a dependency named `environment` could not be found
[2023-04-18T16:31:10Z WARN ] instance `exit` will be imported because a dependency named `exit` could not be found
[2023-04-18T16:31:10Z WARN ] instance `stderr` will be imported because a dependency named `stderr` could not be found
[2023-04-18T16:31:10Z WARN ] instance `streams` will be imported because a dependency named `streams` could not be found
[2023-04-18T16:31:10Z WARN ] instance `filesystem` will be imported because a dependency named `filesystem` could not be found
[2023-04-18T16:31:10Z WARN ] instance `environment` will be imported because a dependency named `environment` could not be found
[2023-04-18T16:31:10Z WARN ] instance `exit` will be imported because a dependency named `exit` could not be found
[2023-04-18T16:31:10Z WARN ] instance `stderr` will be imported because a dependency named `stderr` could not be found
```

When curling the running server, a `500` is returned with error: `import "streams" not defined`.

**Comment by peterhuene** [comment]
Ah, that's a simple matter of updating the server to use the [prototype preview2 host implementation](https://github.com/bytecodealliance/preview2-prototyping/tree/main/host).

**Comment by fibonacci1729** [comment]
Ah thanks Peter! I figured as much! Are we ok with me adding that? I'll drop the target changes and include preview2-prototyping host.

**Comment by peterhuene** [comment]
@fibonacci1729 that'd be great!

**Comment by peterhuene** [comment]
Currently the wasi adapter used by cargo component is `408f0bf`, so we'd need to use a corresponding host.

I need a release of `wit-bindgen` to the latest out of wasm-tools to fully update cargo-component to the latest everything, if we want to wait on that.

**Comment by fibonacci1729** [comment]
cc @peterhuene I made some changes to use the latest `cargo-component` and `preview2-prototyping`. Caveat: I'm not entirely confident this is 100% correct so please suggest changes or improvements!

File: crates/wasm-compose/example/README.md [hunk 1]
```diff
@@ -99,14 +99,14 @@ To build the `service` component, use `cargo component build`:
 
 ```sh
 cd service
-cargo component build --release
+cargo component build --release --target wasm32-wasi
```
Comment (peterhuene) [inline crates/wasm-compose/example/README.md:102]: The default target should already be `wasm32-wasi`.

```suggestion
cargo component build --release
```

File: crates/wasm-compose/example/README.md [hunk 2]
```diff
@@ -99,14 +99,14 @@ To build the `service` component, use `cargo component build`:
...
 ```
 
 To build the `middleware` component, use `cargo component build`:
 
 ```sh
 cd middleware
-cargo component build --release
+cargo component build --release --target wasm32-wasi
```
Comment (peterhuene) [inline crates/wasm-compose/example/README.md:109]:
```suggestion
cargo component build --release
```

File: crates/wasm-compose/example/middleware/Cargo.toml [hunk 3]
```diff
@@ -6,7 +6,7 @@ publish = false
 
 [dependencies]
 flate2 = "1.0.24"
-wit-bindgen = { version = "0.3.0", default_features = false }
+wit-bindgen = { version = "0.4.0", default_features = false }
```
Comment (peterhuene) [inline crates/wasm-compose/example/middleware/Cargo.toml:9]:
```suggestion
wit-bindgen = { version = "0.5.0", default_features = false }
```

File: crates/wasm-compose/example/server/Cargo.toml [hunk 4]
```diff
@@ -9,7 +9,8 @@ async-std = { version = "1.12.0", features = ["attributes"] }
 clap = { version = "3.2.23", features = ["derive"] }
 driftwood = "0.0.6"
 tide = "0.16.0"
-wasmtime = { version = "6.0.0", features = ["component-model"] }
-wasmtime-component-macro = "6.0.0"
+wasi-cap-std-sync = { git = "https://github.com/bytecodealliance/preview2-prototyping" }
```
Comment (peterhuene) [inline crates/wasm-compose/example/server/Cargo.toml:12]: If we want, we can pin this and `wasi-host` to `0d5ca07` to match that we document which `cargo-component` version we tested with.

File: crates/wasm-compose/example/service/Cargo.toml [hunk 5]
```diff
@@ -5,7 +5,7 @@ edition = "2021"
 publish = false
 
 [dependencies]
-wit-bindgen = { version = "0.3.0", default_features = false }
+wit-bindgen = { version = "0.4.0", default_features = false }
```
Comment (peterhuene) [inline crates/wasm-compose/example/service/Cargo.toml:8]:
```suggestion
wit-bindgen = { version = "0.5.0", default_features = false }
```

**Review by peterhuene — COMMENTED** [review]
Looks great! Thanks for updating this. Just a few minor comments.

File: crates/wasm-compose/example/README.md (hunk 1 above)
Comment (fibonacci1729) [reply crates/wasm-compose/example/README.md:102]: Doh! I completely overlooked that. Thanks

**Review by peterhuene — APPROVED** [review]

**Pull request merged**